api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -0,0 +1,437 @@
+        self.i = 0
+        self.stats.n_obs += self.X_batch.shape[0]
+        self.stats.nnz += self.X_batch.nnz
+        self.stats.elapsed += int(time() - start_time)
+        self.stats.n_soma_batches += 1
+        pytorch_logger.debug(f"Retrieved batch: shape={self.X_batch.shape}, cum_stats: {self.stats}")
+        return self.obs_batch_
+
+
+class ExperimentDataPipe(pipes.IterDataPipe[Dataset[ObsDatum]]):  # type: ignore
+    """
+    An iterable-style data pipe.
+    """
+
+    _query: Optional[soma.ExperimentAxisQuery]
+
+    _obs_joinids_partitioned: Optional[pa.Array]
+
+    _obs_and_x_iter: Optional[_ObsAndXIterator]
+
+    _encoders: Optional[Encoders]
+
+    _stats: Stats
+
+    def __init__(
+        self,
+        experiment: soma.Experiment,
+        ms_name: str,
+        layer_name: str,
+        obs_query: Optional[soma.AxisQuery] = None,
+        var_query: Optional[soma.AxisQuery] = None,
+        obs_column_names: Sequence[str] = (),
+        batch_size: int = 1,
+        sparse_X: bool = False,
+        num_workers: int = 0,
+        buffer_bytes: int = DEFAULT_BUFFER_BYTES,
+    ) -> None:
+        if num_workers > 1 and sparse_X:
+            raise NotImplementedError(
+                "torch does not work with sparse tensors in multi-processing mode "
+                "(see https://github.com/pytorch/pytorch/issues/20248)"
+            )
+
+        self.exp_uri = experiment.uri
+        self.aws_region = experiment.context.tiledb_ctx.config().get("vfs.s3.region")
+        self.ms_name = ms_name
+        self.layer_name = layer_name
+        self.obs_query = obs_query
+        self.var_query = var_query
+        self.obs_column_names = obs_column_names
+        self.batch_size = batch_size
+        self.sparse_X = sparse_X
+        # TODO: This will control the obs SOMA batch sizes, and should be replaced with a row count once TileDB-SOMA
+        #  supports `batch_size` param. Unfortunately, the buffer_bytes also impacts the X data fetch efficiency, which
+        #  should be tuned independently.
+        self.buffer_bytes = buffer_bytes
+        self._query = None
+        self._stats = Stats()
+        self._encoders = None
+
+        if "soma_joinid" not in self.obs_column_names:
+            self.obs_column_names = ["soma_joinid", *self.obs_column_names]
+
+    def _init(self) -> None:
+        if self._query is not None:
+            return
+
+        # TODO: support multiple layers, per somacore.query.query.ExperimentAxisQuery.to_anndata()
+        # TODO: for dev-time use, specify smaller batch size via platform_config, once supported
+
+        # TODO: handle closing of exp when iterator is no longer in use; may need be used as a ContextManager,
+        #  but not clear how we can do that when used by DataLoader
+        exp = _open_experiment(self.exp_uri, self.aws_region, buffer_bytes=self.buffer_bytes)
+
+        self._query = exp.axis_query(
```

Comment:
both experiment and query have a thread pool which may leak if not closed.  Query tries to avoid this via __del__ but that is unreliable.  if there is any cleanup protocol with pytorch, would be good to hook into it....

Reply:
Thanks, yeah, I've updated an earlier TODO to capture this: 
```
# TODO: handle closing of `_query` (and transitively, `exp`) when iterator is no longer in use; may need be used as a ContextManager, 
# but not clear how we can do that when used by DataLoader
```